File: crates/lunarbase-client-core/src/transport/ws.rs
//! Generic Ethereum JSON-RPC WebSocket source.
//!
//! The HTTP endpoint remains the authority for block-tagged snapshots and
//! canonical backfill.  This module only consumes the low-latency `logs` and
//! `newHeads` subscriptions and normalizes them into the same source model as
//! the network-specific adapters.  A closed socket is deliberately terminal:
//! callers must recover from the HTTP source before claiming freshness again.

use crate::source::{ChainDataSource, SourceStream};
use crate::state::ordering::CursorReorderBuffer;
use crate::transport::rpc::backend::RpcHttpBackend;
use crate::transport::rpc::client::RpcHttpClient;
use crate::transport::rpc::codec::parse_rpc_log;
use crate::transport::rpc::snapshot::RpcSnapshotProvider;
use crate::{
    bootstrap::BootstrapSnapshot,
    model::{
        BackfillRequest, ChainCursor, ChainUpdate, Checkpoint, Commitment, ContractFilter,
        ContractLog, DeploymentConfig, Network, SourceError,
    },
};
use async_stream::stream;
use futures_util::StreamExt;
use serde_json::Value;
use std::sync::Arc;

mod connection;
mod protocol;

use connection::{establish, websocket_payload};
use protocol::{WsHead, head_discontinuity, parse_ws_head};

/// Bounds transport frames and the number of updates that may wait for a
/// block-head watermark.  Both bounds are required for predictable memory
/// use when an RPC provider stops producing heads or sends a burst of logs.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct WsRpcConfig {
    /// Maximum accepted WebSocket frame size before the stream fails closed.
    pub max_frame_bytes: usize,
    /// Maximum normalized updates retained while waiting for an ordering watermark.
    pub reorder_capacity: usize,
    /// Ethereum subscription method, either `logs` or provider-specific `pendingLogs`.
    pub logs_subscription: String,
    /// Accept multiple monotonically sequenced heads at one block height.
    pub progressive_heads: bool,
}

impl Default for WsRpcConfig {
    fn default() -> Self {
        Self {
            max_frame_bytes: 256 * 1024,
            reorder_capacity: 4096,
            logs_subscription: "logs".into(),
            progressive_heads: false,
        }
    }
}

impl WsRpcConfig {
    /// Validates the hard bounds used to protect the WebSocket ingestion
    /// queue. A zero bound would either make every frame invalid or make the
    /// reorder buffer unable to accept its first update.
    pub fn validate(&self) -> Result<(), SourceError> {
        if self.max_frame_bytes == 0
            || self.reorder_capacity == 0
            || !matches!(self.logs_subscription.as_str(), "logs" | "pendingLogs")
        {
            return Err(SourceError::Unavailable(
                "WS frame and reorder bounds must be non-zero".into(),
            ));
        }
        Ok(())
    }
}

/// Standard JSON-RPC WebSocket backend. It is suitable for an executed local
/// Nitro node and for providers that expose ordinary Ethereum subscriptions.
/// Base Flashblocks providers can use the specialized adapter in
/// `flashblocks.rs` while retaining this backend for canonical fallback.
#[derive(Clone)]
pub struct WsRpcBackend {
    /// Canonical HTTP backend used for snapshots, backfills, and validation.
    http: RpcHttpBackend,
    /// WebSocket endpoint used exclusively for realtime subscriptions.
    ws_endpoint: Arc<str>,
    /// Frame, ordering, and subscription behavior limits.
    config: WsRpcConfig,
}

impl WsRpcBackend {
    /// Creates a standard Ethereum WebSocket backend with conservative
    /// defaults for frame size and out-of-order buffering.
    pub fn new(
        rpc: RpcHttpClient,
        ws_endpoint: impl Into<String>,
        network: Network,
        chain_id: u64,
        snapshot_tag: impl Into<String>,
    ) -> Self {
        Self::with_config(
            rpc,
            ws_endpoint,
            network,
            chain_id,
            snapshot_tag,
            WsRpcConfig::default(),
        )
    }

    /// Creates a WebSocket backend with explicit resource limits.
    ///
    /// The HTTP client remains the source of block-tagged snapshots and
    /// canonical backfills; `ws_endpoint` is used only for realtime logs and
    /// head notifications.
    pub fn with_config(
        rpc: RpcHttpClient,
        ws_endpoint: impl Into<String>,
        network: Network,
        chain_id: u64,
        snapshot_tag: impl Into<String>,
        config: WsRpcConfig,
    ) -> Self {
        Self {
            http: RpcHttpBackend::new(rpc, network, chain_id, snapshot_tag),
            ws_endpoint: Arc::from(ws_endpoint.into()),
            config,
        }
    }

    /// Returns the configured WebSocket endpoint.
    pub fn endpoint(&self) -> &str {
        &self.ws_endpoint
    }

    /// Returns the transport limits used by this backend.
    pub fn config(&self) -> &WsRpcConfig {
        &self.config
    }
}

impl ChainDataSource for WsRpcBackend {
    fn network(&self) -> Network {
        self.http.network()
    }

    async fn snapshot(
        &self,
        deployment: &DeploymentConfig,
    ) -> Result<BootstrapSnapshot, SourceError> {
        RpcSnapshotProvider::new(self.http.rpc().clone(), self.http.snapshot_tag().to_owned())
            .snapshot(deployment)
            .await
    }

    async fn backfill(&self, request: BackfillRequest) -> Result<Vec<ContractLog>, SourceError> {
        self.http.backfill(request).await
    }

    async fn subscribe(&self, filter: ContractFilter) -> Result<SourceStream, SourceError> {
        self.config.validate()?;
        let established = establish(
            self.ws_endpoint.as_ref(),
            &filter,
            &self.config.logs_subscription,
            self.http.chain_id(),
            self.config.max_frame_bytes,
        )
        .await?;
        let logs_subscription = established.logs_subscription;
        let heads_subscription = established.heads_subscription;
        let mut buffered = established.buffered;
        let socket = established.socket;
        let (mut writer, mut reader) = socket.split();

        let chain_id = self.http.chain_id();
        let config = self.config.clone();
        let stream = stream! {
            let mut reorder = match CursorReorderBuffer::new(config.reorder_capacity) {
                Ok(buffer) => buffer,
                Err(error) => {
                    yield Err(error);
                    return;
                }
            };
            let mut last_head: Option<WsHead> = None;
            let mut source_sequence = 0_u64;

            loop {
                let payload = if let Some(payload) = buffered.pop_front() {
                    payload
                } else {
                    let Some(message) = reader.next().await else {
                        yield Ok(ChainUpdate::Gap {
                            cursor: last_head.as_ref().map(|head| head.cursor.clone()),
                            reason: "RPC WebSocket closed; canonical recovery required".into(),
                        });
                        break;
                    };
                    let message = match message {
                        Ok(message) => message,
                        Err(error) => {
                            yield Ok(ChainUpdate::Gap {
                                cursor: last_head.as_ref().map(|head| head.cursor.clone()),
                                reason: format!("RPC WebSocket read failed; canonical recovery required: {error}"),
                            });
                            break;
                        }
                    };
                    match websocket_payload(message, &mut writer).await {
                        Ok(Some(payload)) => payload,
                        Ok(None) => continue,
                        Err(error) => {
                            yield Ok(ChainUpdate::Gap {
                                cursor: last_head.as_ref().map(|head| head.cursor.clone()),
                                reason: error.to_string(),
                            });
                            break;
                        }
                    }
                };
                if payload.len() > config.max_frame_bytes {
                    yield Ok(ChainUpdate::Gap {
                        cursor: last_head.as_ref().map(|head| head.cursor.clone()),
                        reason: "RPC WebSocket frame exceeded configured bound".into(),
                    });
                    break;
                }
                let value: Value = match serde_json::from_slice(&payload) {
                    Ok(value) => value,
                    Err(error) => {
                        yield Ok(ChainUpdate::Gap {
                            cursor: last_head.as_ref().map(|head| head.cursor.clone()),
                            reason: format!("invalid RPC WebSocket JSON; canonical recovery required: {error}"),
                        });
                        break;
                    }
                };

                if let Some(error) = value.get("error") {
                    yield Err(SourceError::Unavailable(format!("RPC subscription error: {error}")));
                    break;
                }
                if value.get("id").is_some() {
                    continue;
                }
                let Some(params) = value.get("params").and_then(Value::as_object) else {
                    continue;
                };
                if value.get("method").and_then(Value::as_str) != Some("eth_subscription") {
                    continue;
                }
                let Some(subscription) = params.get("subscription").and_then(Value::as_str) else {
                    yield Ok(ChainUpdate::Gap {
                        cursor: last_head.as_ref().map(|head| head.cursor.clone()),
                        reason: "RPC subscription notification has no subscription id".into(),
                    });
                    break;
                };
                let Some(result) = params.get("result") else {
                    yield Ok(ChainUpdate::Gap {
                        cursor: last_head.as_ref().map(|head| head.cursor.clone()),
                        reason: "RPC subscription notification has no result".into(),
                    });
                    break;
                };

                if logs_subscription == subscription {
                    let mut log = match parse_rpc_log(result, chain_id, Commitment::Realtime) {
                        Ok(log) => log,
                        Err(error) => {
                            yield Err(SourceError::Unavailable(format!("invalid RPC log notification: {error}")));
                            break;
                        }
                    };
                    source_sequence = source_sequence.saturating_add(1);
                    log.cursor.source_sequence = Some(source_sequence);
                    if let Some(head) = last_head.as_ref()
                        && head.cursor.block_number == log.cursor.block_number
                    {
                        log.cursor.execution_block_number =
                            head.cursor.execution_block_number;
                    }
                    let update = ChainUpdate::Log(log);
                    match reorder.push(update) {
                        Ok(_) => {}
                        Err(error) => {
                            yield Err(error);
                            break;
                        }
                    }
                    if let Some(head) = last_head.as_ref() {
                        for update in reorder.drain_through(&head.cursor) {
                            yield Ok(with_execution_context(update, &head.cursor));
                        }
                    }
                    continue;
                }
                if heads_subscription == subscription {
                    let mut head = match parse_ws_head(result, chain_id) {
                        Ok(head) => head,
                        Err(error) => {
                            yield Err(SourceError::Unavailable(format!("invalid RPC head notification: {error}")));
                            break;
                        }
                    };
                    source_sequence = source_sequence.saturating_add(1);
                    head.cursor.source_sequence = Some(source_sequence);
                    if let Some(previous) = last_head.as_ref() {
                        if head.cursor.block_number
                            > previous.cursor.block_number.saturating_add(1)
                        {
                            yield Ok(ChainUpdate::Gap {
                                cursor: Some(previous.cursor.clone()),
                                reason: "RPC WebSocket skipped one or more block heads; canonical recovery required".into(),
                            });
                            break;
                        }
                        if head_discontinuity(previous, &head, config.progressive_heads) {
                            yield Ok(ChainUpdate::Reorg {
                                old_head: previous.cursor.clone(),
                                new_head: head.cursor.clone(),
                            });
                            reorder = match CursorReorderBuffer::new(config.reorder_capacity) {
                                Ok(buffer) => buffer,
                                Err(error) => {
                                    yield Err(error);
                                    break;
                                }
                            };
                        }
                    }
                    last_head = Some(head.clone());
                    if let Err(error) = reorder.push(ChainUpdate::Head(head.cursor.clone())) {
                        yield Err(error);
                        break;
                    }
                    for update in reorder.drain_through(&head.cursor) {
                        yield Ok(with_execution_context(update, &head.cursor));
                    }
                }
            }
        };
        Ok(Box::pin(stream))
    }

    async fn canonical_head(&self) -> Result<ChainCursor, SourceError> {
        self.http.snapshot_cursor(self.http.network()).await
    }

    async fn validate_checkpoint(&self, checkpoint: &Checkpoint) -> Result<bool, SourceError> {
        self.http.validate_checkpoint(checkpoint).await
    }
}

fn with_execution_context(mut update: ChainUpdate, head: &ChainCursor) -> ChainUpdate {
    if let ChainUpdate::Log(log) = &mut update
        && log.cursor.block_number == head.block_number
    {
        log.cursor.execution_block_number = head.execution_block_number;
    }
    update
}

#[cfg(test)]
#[path = "ws_tests.rs"]
mod tests;

File: packages/client-core/src/transport/ws.ts
import type {
  BackfillRequest,
  BootstrapSnapshot,
  ChainCursor,
  ChainDataSource,
  ChainUpdate,
  Checkpoint,
  ContractFilter,
  ContractLog,
  DeploymentConfig,
  Network,
} from "../model.js";
import type { Hex } from "ox/Hex";
import { Commitment } from "../model.js";
import { CursorReorderBuffer } from "../state/ordering.js";
import { JsonRpcHttpClient, parseRpcLog, RpcError, RpcHttpBackend, RpcSnapshotProvider } from "./rpc.js";
import {
  defaultWebSocketFactory,
  establishSocket,
  type EstablishedSocket,
  type WebSocketFactory,
} from "./ws/connection.js";
import { gap, parseHead } from "./ws/protocol.js";

export { BoundedFrameQueue, defaultWebSocketFactory } from "./ws/connection.js";
export type { SocketEvent, WebSocketFactory, WebSocketLike } from "./ws/connection.js";

/** Resource bounds for generic Ethereum WebSocket ingestion. */
export interface WsRpcConfig {
  /** Maximum accepted WebSocket frame size before fail-closed recovery. */
  readonly maxFrameBytes: number;
  /** Maximum updates retained while awaiting an ordering watermark. */
  readonly reorderCapacity: number;
  /** Maximum decoded frames waiting for the socket consumer. */
  readonly queueCapacity: number;
  /** Ethereum subscription method used for Core event logs. */
  readonly logsSubscription: "logs" | "pendingLogs";
  /** Whether multiple sequenced heads at one block height are valid progress. */
  readonly progressiveHeads: boolean;
}

export const DEFAULT_WS_RPC_CONFIG: WsRpcConfig = Object.freeze({
  maxFrameBytes: 256 * 1024,
  reorderCapacity: 4096,
  queueCapacity: 4096,
  logsSubscription: "logs",
  progressiveHeads: false,
});

/**
 * Standard Ethereum JSON-RPC WebSocket source. HTTP remains authoritative
 * for block-tagged bootstrap and canonical backfill; a socket gap is emitted
 * instead of being silently hidden by reconnecting from an unknown cursor.
 */
export class WsRpcBackend implements ChainDataSource {
  /** Canonical HTTP backend used for heads, backfill, and validation. */
  private readonly http: RpcHttpBackend;
  /** Coherent block-tagged Core snapshot provider. */
  private readonly snapshots: RpcSnapshotProvider;
  /** Platform or injected WebSocket constructor. */
  private readonly factory: WebSocketFactory;
  /** Validated frame, queue, ordering, and subscription limits. */
  readonly config: WsRpcConfig;

  /** Creates a WebSocket backend with bounded frame, queue, and reorder memory. */
  constructor(
    /** Strict read-only HTTP client used outside the quote path. */
    readonly rpc: JsonRpcHttpClient,
    /** WebSocket endpoint used exclusively for realtime subscriptions. */
    readonly wsEndpoint: string,
    /** Network family exposed through the common source interface. */
    readonly network: Network,
    /** EIP-155 chain identifier attached to normalized cursors. */
    readonly chainId: bigint,
    /** Canonical block tag used for snapshot and recovery reads. */
    readonly snapshotTag = "latest",
    config: Partial<WsRpcConfig> = {},
    factory: WebSocketFactory = defaultWebSocketFactory,
  ) {
    this.config = validateConfig({ ...DEFAULT_WS_RPC_CONFIG, ...config });
    this.factory = factory;
    this.http = new RpcHttpBackend(rpc, network, chainId, snapshotTag);
    this.snapshots = new RpcSnapshotProvider(rpc, snapshotTag);
  }

  /** Reads one coherent quote state through block-tagged HTTP calls. */
  snapshot(deployment: DeploymentConfig): Promise<BootstrapSnapshot> {
    if (deployment.network !== this.network)
      return Promise.reject(new RpcError("INVALID", "RPC source network mismatch"));
    return this.snapshots.snapshot(deployment);
  }

  /** Delegates canonical log backfill to HTTP. */
  backfill(request: BackfillRequest): Promise<readonly ContractLog[]> {
    return this.http.backfill(request);
  }

  /**
   * Opens and acknowledges logs/new-head subscriptions before exposing the
   * stream. A client therefore cannot become ready while the socket is merely
   * connecting or while the node has rejected either subscription.
   */
  async subscribe(filter: ContractFilter, signal?: AbortSignal): Promise<AsyncIterable<ChainUpdate>> {
    const connection = await establishSocket(
      this.wsEndpoint,
      this.factory,
      filter,
      this.config.logsSubscription,
      this.chainId,
      this.config.queueCapacity,
      this.config.maxFrameBytes,
      signal,
    );
    return this.readSocket(connection, signal);
  }

  /** Returns the canonical HTTP recovery head. */
  canonicalHead(): Promise<ChainCursor> {
    return this.http.canonicalHead();
  }

  /** Verifies a checkpoint block hash through canonical HTTP RPC. */
  validateCheckpoint(checkpoint: Checkpoint): Promise<boolean> {
    return this.http.validateCheckpoint(checkpoint);
  }

  private async *readSocket(connection: EstablishedSocket, signal?: AbortSignal): AsyncIterable<ChainUpdate> {
    const { queue, logsSubscription, headsSubscription, prefetched, close } = connection;
    let lastHead: ChainCursor | undefined;
    let lastParentHash: Hex | undefined;
    let sourceSequence = 0n;
    let reorder = new CursorReorderBuffer(this.config.reorderCapacity);

    try {
      while (!signal?.aborted) {
        let frame: string | undefined;
        try {
          frame = prefetched.shift() ?? (await queue.next(signal));
        } catch (error) {
          yield gap(`RPC WebSocket failed; canonical recovery required: ${message(error)}`, lastHead);
          return;
        }
        if (frame === undefined) {
          yield gap("RPC WebSocket closed; canonical recovery required", lastHead);
          return;
        }
        let value: Record<string, unknown>;
        try {
          value = JSON.parse(frame) as Record<string, unknown>;
        } catch (error) {
          yield gap(`invalid RPC WebSocket JSON; canonical recovery required: ${message(error)}`, lastHead);
          return;
        }
        if (value.error) {
          yield gap(`RPC subscription error: ${JSON.stringify(value.error)}`, lastHead);
          return;
        }
        if (value.method !== "eth_subscription" || !value.params || typeof value.params !== "object") continue;
        const params = value.params as Record<string, unknown>;
        const subscription = typeof params.subscription === "string" ? params.subscription : undefined;
        if (!subscription || params.result === undefined) {
          yield gap("RPC subscription notification is missing subscription/result", lastHead);
          return;
        }

        if (subscription === logsSubscription) {
          let log: ContractLog;
          try {
            log = parseRpcLog(params.result, this.chainId, Commitment.Realtime);
          } catch (error) {
            yield gap(`invalid RPC log notification: ${message(error)}`, lastHead);
            return;
          }
          sourceSequence += 1n;
          log = { ...log, cursor: { ...log.cursor, sourceSequence } };
          if (lastHead?.blockNumber === log.cursor.blockNumber)
            log = withExecutionContext(log, lastHead.executionBlockNumber);
          try {
            reorder.push({ kind: "Log", log });
          } catch (error) {
            yield gap(`RPC reorder buffer failed: ${message(error)}`, lastHead);
            return;
          }
          if (lastHead)
            for (const update of reorder.drainThrough(lastHead)) yield annotateExecutionContext(update, lastHead);
          continue;
        }

        if (subscription !== headsSubscription) continue;
        let parsed: { cursor: ChainCursor; parentHash?: Hex };
        try {
          parsed = parseHead(params.result, this.chainId);
        } catch (error) {
          yield gap(`invalid RPC head notification: ${message(error)}`, lastHead);
          return;
        }
        sourceSequence += 1n;
        parsed.cursor = { ...parsed.cursor, sourceSequence };
        if (lastHead && headDiscontinuity(lastHead, lastParentHash, parsed, this.config.progressiveHeads)) {
          yield { kind: "Reorg", oldHead: lastHead, newHead: parsed.cursor };
          reorder = new CursorReorderBuffer(this.config.reorderCapacity);
        }
        lastHead = parsed.cursor;
        lastParentHash = parsed.parentHash;
        try {
          reorder.push({ kind: "Head", cursor: parsed.cursor });
        } catch (error) {
          yield gap(`RPC reorder buffer failed: ${message(error)}`, lastHead);
          return;
        }
        for (const update of reorder.drainThrough(parsed.cursor)) yield annotateExecutionContext(update, parsed.cursor);
      }
    } finally {
      close();
    }
  }
}

function validateConfig(config: WsRpcConfig): WsRpcConfig {
  for (const [name, value] of Object.entries({
    maxFrameBytes: config.maxFrameBytes,
    reorderCapacity: config.reorderCapacity,
    queueCapacity: config.queueCapacity,
  }))
    if (!Number.isSafeInteger(value) || value <= 0) throw new Error(`WS ${name} must be a positive safe integer`);
  return Object.freeze(config);
}

function headDiscontinuity(
  lastHead: ChainCursor,
  lastParentHash: Hex | undefined,
  next: { cursor: ChainCursor; parentHash?: Hex },
  progressiveHeads: boolean,
): boolean {
  if (next.cursor.blockNumber > lastHead.blockNumber + 1n) return true;
  const sameHeight = next.cursor.blockNumber === lastHead.blockNumber;
  const changedParent =
    next.parentHash !== undefined && lastParentHash !== undefined && next.parentHash !== lastParentHash;
  return (
    next.cursor.blockNumber < lastHead.blockNumber ||
    (sameHeight && (!progressiveHeads || changedParent)) ||
    (next.cursor.blockNumber === lastHead.blockNumber + 1n &&
      next.parentHash !== undefined &&
      lastHead.blockHash !== undefined &&
      next.parentHash !== lastHead.blockHash)
  );
}

function annotateExecutionContext(update: ChainUpdate, head: ChainCursor): ChainUpdate {
  if (update.kind !== "Log" || update.log.cursor.blockNumber !== head.blockNumber) return update;
  return { kind: "Log", log: withExecutionContext(update.log, head.executionBlockNumber) };
}

function withExecutionContext(log: ContractLog, executionBlockNumber: bigint): ContractLog {
  return { ...log, cursor: { ...log.cursor, executionBlockNumber } };
}

function message(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}
